uni/ulsp/hover.icn

```diff
@@ -0,0 +1,206 @@
+                desired_line ? {
+                    tab(many(' \t'))
+                    move(9)
+                    tab(many(' \t'))
+                    procedure_name := tab(many(identifiers))
+                }
+
+                num_params := *workspace.internal_procedures[procedure_name]["params"]
+                every param := workspace.internal_procedures[procedure_name]["params"][1 to num_params] do {
+                    if param["name"] == getHoverItem(line, character) then return "parameter"
+                }
+                break
+            }
+
+            # If item is param of next closest method then return param
+            else if ReMatch("^[\ \t]*method[\ \t]+", desired_line) then {
+                desired_line ? {
+                    tab(many(' \t'))
+                    move(6)
+                    tab(many(' \t'))
+                    method_name := tab(many(identifiers))
+                }
+                line_num -:= 1
+                until line_num = 0 do {
+                    desired_line := getDesiredLine(line_num, workspace.temp_file)
+                    if ReMatch("^[\ \t]*class[\ \t]+", desired_line) then {
+                        desired_line ? {
+                            tab(many(' \t'))
+                            move(5)
+                            tab(many(' \t'))
+                            class_name := tab(many(identifiers))
+                        }
+                        break
+                    }
+                    line_num -:= 1
+                }
+
+                num_params := *workspace.internal_classes[class_name]["methods"][method_name]["params"]
+                every param := workspace.internal_classes[class_name]["methods"][method_name]["params"][1 to num_params] do {
+                    if param["name"] == getHoverItem(line, character) then return "parameter"
+                }
+
+                # If item is an attribute (class parameter) then return class parameter
+                num_attributes := *workspace.internal_classes[class_name]["attributes"]
+                every _attribute := workspace.internal_classes[class_name]["attributes"][1 to num_attributes] do {
+                    if _attribute["name"] == getHoverItem(line, character) then return "class parameter"
+                }
+                break
+            }
+            line_num -:= 1
+        }
+
+        line_num := params["position"]["line"]
+
+        # If item is a member of global variables then return global
+        if member(workspace.global_variables, getHoverItem(line, character)) then return "global"
+
+        # If item is an associated variable or object then return variable or object respectively
+        if member(workspace.objects, getHoverItem(line, character)) then return "object"
+        if member(workspace.variables, getHoverItem(line, character)) then return "variable"
+    end
+
+    method getHoverItem(line, character)
+    local identifiers
```

Comment:
indent.

and to make the initialization be done once and not every time you call the function, you can do 

```
static identifiers
initial identifiers := &letters ++ &digits ++ '_' ++ '-'
```